crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,767 @@
+        goto err;
+
+    if (!(ASN1_BIT_STRING_set(ps->signature, sig, siglen)))
+        goto err;
+
+    /* cleanup */
+    OPENSSL_free(crder);
+    EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return 1;
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+    OPENSSL_free(crder);
+    if (ctx)
+        EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return 0;
+}
+
+
+int OSSL_CRMF_MSG_create_popo(OSSL_CRMF_MSG *crm, const EVP_PKEY *pkey,
+                              int dgst, int ppmtd)
+{
+    OSSL_CRMF_POPO *pp = NULL;
+
+    if (crm == NULL
+        || (ppmtd == OSSL_CRMF_POPO_SIGNATURE && pkey == NULL)) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if ((pp = OSSL_CRMF_POPO_new()) == NULL)
+        goto oom;
+    pp->type = ppmtd;
+
+    switch (ppmtd) {
+    case OSSL_CRMF_POPO_NONE:
+        OSSL_CRMF_POPO_free(pp);
+        OSSL_CRMF_POPO_free(crm->popo);
+        crm->popo = NULL;
+        return 1;
+    case OSSL_CRMF_POPO_RAVERIFIED:
+        if ((pp->value.raVerified = ASN1_NULL_new()) == NULL)
+            goto oom;
+        break;
+
+    case OSSL_CRMF_POPO_SIGNATURE:
+        {
+            OSSL_CRMF_POPOSIGNINGKEY *ps = OSSL_CRMF_POPOSIGNINGKEY_new();
+            if (ps == NULL
+                || !CRMF_poposigningkey_init(ps, crm->certReq, pkey, dgst)) {
+                OSSL_CRMF_POPOSIGNINGKEY_free(ps);
+                goto err;
+            }
+            pp->value.signature = ps;
+        }
+        break;
+
+    case OSSL_CRMF_POPO_KEYENC:
+        if ((pp->value.keyEncipherment = OSSL_CRMF_POPOPRIVKEY_new()) == NULL)
+            goto oom;
+        pp->value.keyEncipherment->type = OSSL_CRMF_POPOPRIVKEY_SUBSEQUENTMESSAGE;
+        if ((pp->value.keyEncipherment->value.subsequentMessage =
+             ASN1_INTEGER_new()) == NULL
+            ||
+            !ASN1_INTEGER_set(pp->value.keyEncipherment->value.subsequentMessage,
+                              OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT))
+            goto oom;
+        break;
+
+    default:
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO,
+                CRMF_R_UNSUPPORTED_METHOD_FOR_CREATING_POPO);
+        goto err;
+    }
+
+    OSSL_CRMF_POPO_free(crm->popo);
+    crm->popo = pp;
+
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO, ERR_R_MALLOC_FAILURE);
+ err:
+    OSSL_CRMF_POPO_free(pp);
+    return 0;
+}
+
+/* returns 0 for equal, -1 for a < b, 1 for a > b, <= -2 on error */
+static int X509_PUBKEY_cmp(X509_PUBKEY *a, X509_PUBKEY *b)
+{
+    X509_ALGOR *algA = NULL, *algB = NULL;
+    int res = 0;
+
+    if (a == b)
+        return 0;
+    if (a == NULL)
+        return -1;
+    if (b == NULL)
+        return 1;
+    if (!X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a)
+        ||
+        !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b))
+        return -2;
+    if (algA == NULL || algB == NULL)
+        return -3;
```

Comment:
I'd don't like the magic number return values. Better would be an enum or macro constants.

Reply:
The whole static function could be simplified.

Reply:
I've simplified the result values of `X509_PUBKEY_cmp()` to the typical choice of {-1, 0, 1}.